docs/hugo/content/design/ADR-2022-11-Change-Detection.md

```diff
@@ -4,79 +4,115 @@ title: '2022-11: Change Detection'
 
 A variation of value defaulting may occur with arrays and maps, where some services will add a default value to the collection if the user does not specify it. Conceptually, you can write `["a"]` and get back `["default", "a"]` when you read it back.
 
-### Read-After-Write Consistency
+**Potential mitigation**: Keep track of the changes made by Azure when we originally PUT the resource, and use those as another input to our change detection process. For most resources, the current state of the resource is returned in the response to either the PUT or the long-running-operation (LRO) polling operation, so we can capture these changes at that point. For others, we may need to perform an initial GET after the PUT to capture these changes, despite the inherent race condition of doing so.
+
+### Issue: Read-After-Write Consistency
 
-Azure is not guaranteed to return the exact same resource that was PUT. 
+Azure is not guaranteed to return the exact same resource that was PUT.
 
 * Some services normalize identifier fields (converting them to all lowercase)
 * Some fields accept a wider range of values than they return.
 
-For example the CosmosDB API will return the region `West US` when the resource specified `westus`.
+For example, the CosmosDB API will return the region `West US` when the resource specified `westus`.
+
+**Potential mitigations**:
+
+* Generally speaking, Azure identifiers are expected to be "case-preserving, case-insensitive", so we should compare those string fields in a case-insensitive manner.
+* Region names and virtual machine SKUs are two known special cases where we may need to implement special case handling.
+* Capturing the changes (as detailed above) would also work for this issue.
 
-Potential mitigations: 
+### Issue: Array ordering
 
-* Generally speaking, Azure identifiers are expected to be "case-preserving, case-insensitive", so we should compare those string fields in a case insensitive manner. 
-* Region names and virtual machines SKUS are two known special cases where we may need to do some special case handling. 
+Some resources have arrays of items. These arrays may not be guaranteed to be returned in the same order as they were specified.
 
-### Array ordering
+**Potential mitigations**:
 
-Some resources have arrays of items. These arrays may not be guaranteed to be returned in the same order as they were specified. 
+* In the common case where the items in an array have a known identifier (e.g., when they are sub-resources with an `id` field), use that identifier to match them up. Where the items in an array do not have a known identifier, compare them by index.
+* Add specific configuration for array properties to specify how they should be compared — whether by an id field or by index. Consistent with our similar configuration used elsewhere, we'd require every array property to be annotated to ensure we don't do the wrong thing by default.
```

Comment:
FWIW Swagger actually has this field. It's called [x-ms-identifiers](https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-identifiers) and it can be used to call out which field to use as the key.

I think it has pretty wide adoption because there's a linter that checks for it. With that said, you can say `[]` in which case we'd have to assume "by index"